firedrake/preconditioners/hypre_ams.py

```diff
@@ -31,14 +32,26 @@ def initialize(self, obj):
         P1 = FunctionSpace(mesh, "Lagrange", 1)
         G = Interpolator(grad(TestFunction(P1)), V).callable().handle
 
+        # remove (near) zeros from sparsity pattern
+        ai, aj, a = G.getValuesCSR()
+        a[np.abs(a) < 1e-10] = 0
+        G2 = PETSc.Mat().create()
+        G2.setType(PETSc.Mat.Type.AIJ)
+        G2.setSizes(G.sizes)
+        G2.setOption(PETSc.Mat.Option.IGNORE_ZERO_ENTRIES, True)
+        G2.setUp()
+        G2.setValuesCSR(ai,aj,a)
```

Comment:
Sorry it's taken so long to get to. I think this all looks good apart from the suggestion below:

```suggestion
        G2.setPreallocationCSR((ai,aj,a))
```

This does the correct preallocation and sets values all in one go.

We want to get the preallocation right because otherwise the assembly might be very slow.